src/Control/Exception/Annotated.hs

```diff
@@ -112,30 +112,40 @@ instance (Exception exception) => Exception (AnnotatedException exception) where
     fromException exn
         | Just (e :: exception) <- Safe.fromException exn
         =
-            pure $ new e
+            pure $ pure e
         | otherwise
         =
             Nothing
 
--- | Attach an empty @['Annotation']@ to an exception.
+-- | Annotate the underlying exception with a 'CallStack'.
 --
--- @since 0.1.0.0
-new :: e -> AnnotatedException e
-new = pure
+-- @since 0.2.0.0
+exceptionWithCallStack :: (Exception e, HasCallStack) => e -> AnnotatedException e
+exceptionWithCallStack =
+    AnnotatedException [callStackAnnotation]
 
 -- | Append the @['Annotation']@ to the 'AnnotatedException'.
 --
+-- 'CallStack' is a special case - if a 'CallStack' is present in both the
+-- 'AnnotatedException' and the @['Annotation']@, then this will append the
+-- 'CallStack's in the new list and concatenate them all together.
+--
 -- @since 0.1.0.0
 annotate :: [Annotation] -> AnnotatedException e -> AnnotatedException e
-annotate ann (AnnotatedException anns e) = AnnotatedException (ann ++ anns) e
+annotate newAnnotations (AnnotatedException oldAnnotations e) =
+    let
+        (callStacks, other) =
+            tryAnnotations (newAnnotations <> oldAnnotations)
+    in
+        foldr addCallStackToException (AnnotatedException other e) callStacks
```

Comment:
Wouldn't it make more sense for `AnnotatedException` to have a `Maybe CallStack` field, instead of extracting out the callstacks and treating them specially?

Reply:
No, it's a great question, and something I waffled on a bit. 

It's *nice* to have the exposed constructor and just write `AnnotatedException annotations blah` and then do `foldr whatever def annotations`. I like that. 

It's also nice to have a notion of uniformity for how annotations are handled. It's a list. I, as a library author, don't get any special tricks that you, a library consumer, don't also have access to. So if I want to treat `CallStack` in a special way, I do so with the same tools and techniques as anyone else would. In that way, the implementation in the library serves as a sort of "how to manual" on *using* the library as well, if you studied the implementation.

`CallStack` seems like a reasonable exception. And it would be easy to have:

```haskell
data Annotations = Annotations { getAnnotations :: [Annotation], annotationsCallStack :: Maybe CallStack }
```

But now, suppose I write this:

```haskell
lmao :: HasCallStack => IO a -> IO a
lmao =
  checkpoint (Annotation callStack)
```

This is going to slam the `CallStack` into `[Annotation]`, not the `Maybe CallStack` field. Unless I *also* do a type-check on all the functions which can add to the `[Annotation]` to defer to the `CallStack` instead, if the type works. Possible, sure:

```haskell
annotate :: [Annotation] -> AnnotatedException e -> AnnotatedException e
annotate anns (AnnotatedException oldAnns e) = AnnotatedException newAnns e
  where
    newAnns = foldr updateAnnotation oldAnns anns
    updateAnnotation ann acc =
        case castAnnotation ann of
            Just callstack ->
                acc { annotationCallStack = mergeCallStack callstack (annotationCallStack acc) }
            Nothing -> 
                acc { getAnnotations = ann : getAnnotations acc }
```

And this logic had to be duplicated everywhere I modify `Annotations`, much like how it is done today.

Making it flexible/extensible would be something like, uh,

```haskell
data TypeMap (xs :: [Type])

data SomeTypeMap where
    SomeTypeMap :: TypeMap xs -> SomeTypeMap

data Annotations = Annotations { regularAnnotations :: [Annotation], specialAnnotations :: SomeTypeMap }

annotate :: Annotation -> Annotations -> Annotations
annotate ann anns = 
    case ann of
        Annotation (a :: a) ->
            case lookupTypeMap @a (specialAnnotations anns) of
                Just existing ->
                    anns { specialAnnotations = mergeAnnotation a (specialAnnotations anns) }
                Nothing ->
                    anns { regularAnnotations = ann : regularAnnotations anns }                     
```

But then, the "special handling of annotations" is dependent on the specific *value* of the annotation in question, so you'd want to define something like:

```haskell
myThrow = throwIO (AnnotatedException myEmptyAnnotations e)

myEmptyAnnotations = Annotations [] what
  where
    what = SomeTypeMap 
      $ TypeMap.handle @CallStack mergeCallStack
      $ TypeMap.handle @Foobar mergeFooBar
```

Which also seems weird - seems *weird* for an exception to know how to handle it's own annotations.

Reply:
I understand your thinking. I guess I disagree in that I'm way more open to the idea of hiding the data constructors, so the `lmao` scenario couldn't happen. Also informing my opinion is the notion that the callstack is not (or at least should not be) a part of the semantics of the Haskell language, so hiding it to _prevent_ people from branching on it would be an ideal feature. But I understand that people can have different goals, so thank you for explaining how your design does, in fact, meet your goals quite well.

Reply:
> I guess I disagree in that I'm way more open to the idea of hiding the data constructors, so the lmao scenario couldn't happen.

Well, supposing I replace `Annotation` with `toAnnotation :: (Typeable a, Show a) => a -> Annotation`, that still doesn't help, because `Annotation` doesn't handle anything specially - `AnnotationList` would.

Another option would be something like:

```haskell
class (Typeable a) => ToAnnotation a where
  toAnnotation :: a -> AnnotationList
  default toAnnotation :: (Show a, Typeable a) => a -> AnnotationList
  toAnnotation a = mempty { annotations = [Annotation a] }

instance ToAnnotation CallStack where
  toAnnotation cs = mempty { annotationCallStack = Just cs }

instance Semigroup AnnotationList where
  Annotations anns mcs <> Annotations anns' mcs' =
    Annotations (anns <> anns') (mergeCallStacks mcs mcs')

instance Monoid AnnotationList where
  mempty = AnnotationList mempty Nothing
```

🤔 

But then you have to write a `ToAnnotation` instance for basically everything, which means you'd need to expose (at least) set/modify on `AnnotationList`, and you'd need the constructor for `Annotation` available. Which gets us back to our original problem of `Annotation callStack` not working how we'd want.